src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala

```diff
@@ -0,0 +1,303 @@
+      .annotateGlobal(globalSg2: Double, TDouble, s"$rootGA.sigmaG2")
+      .annotateGlobal(globalSe2: Double, TDouble, s"$rootGA.sigmaE2")
+      .annotateGlobal(delta: Double, TDouble, s"$rootGA.delta")
+      .annotateGlobal(h2: Double, TDouble, s"$rootGA.h2")
+      .annotateGlobal(S.data.reverse: IndexedSeq[Double], TArray(TDouble), s"$rootGA.evals")
+
+    val vds2 = if (optDelta.isEmpty) {
+      val maxLogLkhd = diagLMM.maxLogLkhd.get
+      val (deltaGrid, logLkhdVals) = diagLMM.deltaGridLogLkhd.get.unzip
+      vds1
+        .annotateGlobal(maxLogLkhd: Double, TDouble, s"$rootGA.maxLogLkhd")
+        .annotateGlobal(deltaGrid: IndexedSeq[Double], TArray(TDouble), s"$rootGA.logDeltaGrid")
+        .annotateGlobal(logLkhdVals: IndexedSeq[Double], TArray(TDouble), s"$rootGA.logLkhdVals")
+    } else
+      vds1
+
+    if (noAssoc)
+      vds2
+    else {
+      val (newVAS, inserter) = vds.insertVA(LinearMixedModel.schema, pathVA)
+
+      vds2.mapAnnotations { case (v, va, gs) =>
+        val lmmregStat =
+          if (keepAll || assocVariantFilterPred(v, va, gs)) {
+            val SparseGtVectorAndStats(x0, isConstant, maf, nHomRef, nHet, nHomVar, nMissing) = {
+              val sb = new SparseGtBuilder()
+              gs.iterator.zipWithIndex.foreach { case (g, i) => if (sampleMaskBc.value(i)) sb.merge(g) }
+              sb.toSparseGtVector(n)
+            }
+
+            if (!isConstant) {
+              val x: Vector[Double] = if (maf <= sparsityThreshold) x0 else x0.toDenseVector
+
+              val (b, s2, chi2, p) = scalerLMMBc.value.likelihoodRatioTest(TBc.value * x)
+
+              Some(Annotation(b, s2, chi2, p, maf, nHomRef, nHet, nHomVar, nMissing))
+            } else
+              None
+          } else
+            None
+
+        inserter(va, lmmregStat)
+      }.copy(vaSignature = newVAS)
+    }
+  }
+}
+
+object DiagLMM {
+  def apply(
+    C: DenseMatrix[Double],
+    y: DenseVector[Double],
+    S: DenseVector[Double],
+    optDelta: Option[Double] = None,
+    useML: Boolean = false): DiagLMM = {
+
+    require(C.rows == y.length)
+
+    val (delta, maxLogLkhd, gridLogLkhd) =
+      optDelta match {
+        case Some(d) => (d, None, None)
+        case None => fitDelta(C, y, S, useML)
+      }
+
+    val n = y.length
+    val sqrtInvD = sqrt(S + delta).map(1 / _)
+    val TC = C(::, *) :* sqrtInvD
+    val Ty = y :* sqrtInvD
+    val TyTy = Ty dot Ty
+    val TCTy = TC.t * Ty
+    val TCTC = TC.t * TC
+    val b = TCTC \ TCTy
+    val s2 = (TyTy - (TCTy dot b)) / (if (useML) n else n - C.cols)
+
+    DiagLMM(b, s2, math.log(s2), delta, maxLogLkhd, gridLogLkhd, sqrtInvD, TC, Ty, TyTy, useML)
+  }
+
+  def fitDelta(C: DenseMatrix[Double], y: DenseVector[Double], S: DenseVector[Double], useML: Boolean): (Double, Option[Double], Option[IndexedSeq[(Double, Double)]]) = {
+
+    val n = y.length
+    val c = C.cols
+    val shift =
```

Comment:
Why not put shift inside the ML and REML functions?  Then you won't need the conditional.